30066 - Migrate to Gotenberg

Jxio opened the pull request:
https://github.com/bcgov/entity/issues/30066

seeker25 commented on report-api/src/api/services/chunk_report_service.py:
```diff
@@ -74,35 +72,38 @@ def _merge_pdf_files(temp_files: List[str]) -> bytes:
-        return pdf_content
+        gotenberg_url = current_app.config.get('GOTENBERG_URL', 'http://localhost:3000')
+
+        # Prepare the HTML content for Gotenberg
+        html_data = html_out.encode('utf-8')
+
+        # Create multipart form data for Gotenberg
+        data = aiohttp.FormData()
+        data.add_field('index.html', html_data, filename='index.html', content_type='text/html')
+
+        # Call Gotenberg using the shared session
```
comment

seeker25 commented on report-api/src/api/services/chunk_report_service.py:
```diff
@@ -74,35 +72,38 @@ def _merge_pdf_files(temp_files: List[str]) -> bytes:
+        async with session.post(
+            f'{gotenberg_url}/forms/chromium/convert/html',
+            data=data,
+            timeout=aiohttp.ClientTimeout(total=500)  # 500 second timeout
```
comment

seeker25 commented on report-api/src/api/services/chunk_report_service.py:
```diff
@@ -159,77 +161,144 @@ def _prepare_chunk_tasks(
+                async_tasks.append(task)
+                order_ids.append(oid)
+
+            # Run all tasks concurrently using gather
```
comment

seeker25 commented on report-api/src/api/services/chunk_report_service.py:
```diff
@@ -159,77 +161,144 @@ def _prepare_chunk_tasks(
+        try:
+            with pikepdf.Pdf.open(io.BytesIO(pdf_bytes)) as pdf:
+                for i, page in enumerate(pdf.pages):
+                    # create new PDF for each page
```
comment

seeker25 commented on report-api/src/api/services/chunk_report_service.py:
```diff
@@ -159,77 +161,144 @@ def _prepare_chunk_tasks(
-    @dataclass
-    class MergeContext:
-        """Merge context for chunk report."""
+                    # save to bytes
```
comment

seeker25 commented on report-api/src/api/services/chunk_report_service.py:
```diff
@@ -74,35 +72,38 @@ def _merge_pdf_files(temp_files: List[str]) -> bytes:
+        # Prepare the HTML content for Gotenberg
+        html_data = html_out.encode('utf-8')
+
+        # Create multipart form data for Gotenberg
```
comment

seeker25 commented on report-api/src/api/services/chunk_report_service.py:
```diff
@@ -253,40 +322,47 @@ def create_chunk_report(
-        # First pass: quick render to PDF bytes and collect page counts
-        first_pass_pdfs = ChunkReportService._render_tasks_parallel(
-            tasks, base_url, worker_count
+        # First pass: render all chunks to PDF bytes in parallel (no footers)
```
good comment

ochiu commented on report-api/src/api/services/report_service.py:
```diff
@@ -65,14 +64,66 @@ def _finalize_pdf(
                 generate_page_number,
             )
-        return ReportService.generate_pdf(html_out, generate_page_number)
+        return ReportService.generate_pdf(html_out, generate_page_number, footer_html=footer_html)
```
Looks like footer_html is none in L67.

The function also defaults to None, so we can just remove footer_html variable here.

ochiu reviewed (approved):
Great work @Jxio !!

Jxio commented:
Some updates after review:
<img width="821" height="114" alt="image" src="https://github.com/user-attachments/assets/5f8b4423-fa67-4235-bd10-82b48a5cef0e" />

1. no footer start from 2nd page when report page number > 500

2. some code refactor:

- I added gotenberg_service.py, for all gotenberg relate services
- added footer_service.py, for footer rendering, cause it's generic for all of reports, it was in chunk_report_service.py, but chunk_report_service just specific for statement.
